### PR TITLE
Change git.cmd to RunWithContext

### PR DESCRIPTION
**Pull request by martinscholz83**
Change all `cmd...Pipeline` commands from  to `cmd.RunWithContext`.
#18553

### CODE REVIEW

**Comment by 6543** [comment]
Can you also delete tje ...pipeline... functions?

**Comment by martinscholz83** [comment]
You mean the `PipelineFunc` member of the `RunContext` struct?

**Review by zeripath — APPROVED** [review]

**Comment by lunny** [comment]
Maybe you need add `Timeout: -1` for those changes.

**Comment by martinscholz83** [comment]
👍 Wasn't sure about the `Timeout`. Just saw now that these pipeline functions have different overloads to set `Timeout` to `-1`. @6543, you mean deleting the Pipeline methods for the command struct like `func (c *Command) RunInDirTimeoutEnvFullPipeline`, right?

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
Timeout is less of a problem now that the process go context is passed down to the git command.

TBH we should probably be moving the go context into the RunContext e.g. ```go
// RunContext represents parameters to run the command
type RunContext struct {
        Context
	Env            []string
	Timeout        time.Duration
	Dir            string
	Stdout, Stderr io.Writer
	Stdin          io.Reader
	PipelineFunc   func(context.Context, context.CancelFunc) error
}
```

and dropping the parent context stuff from the Command.

**Comment by 6543** [comment]
> 👍 Wasn't sure about the `Timeout`. Just saw now that these pipeline functions have different overloads to set `Timeout` to `-1`. @6543, you mean deleting the Pipeline methods for the command struct like `func (c *Command) RunInDirTimeoutEnvFullPipeline`, right?

Yes

**Comment by martinscholz83** [comment]
I think i missed some places like [this](https://github.com/go-gitea/gitea/blob/main/modules/git/pipeline/catfile.go#L59)

**Comment by lunny** [comment]
> I think i missed some places like [this](https://github.com/go-gitea/gitea/blob/main/modules/git/pipeline/catfile.go#L59)

How about to delete the method `RunInDirPipeline` if all of it have been replaced?

**Comment by martinscholz83** [comment]
While deleting the `RunInDirPipeline` I realized that these `RunInDir` also need to replaced with `RunWithContext`. Is this correct? Like [here](https://github.com/go-gitea/gitea/blob/main/services/mirror/mirror_pull.go#L56). So the only method `command`should implement is `RunWithContext`?

**Comment by 6543** [comment]
yes there are different RinIn ... func that all can be replaced

**Comment by lunny** [comment]
> While deleting the `RunInDirPipeline` I realized that these `RunInDir` also need to replaced with `RunWithContext`. Is this correct? Like [here](https://github.com/go-gitea/gitea/blob/main/services/mirror/mirror_pull.go#L56). So the only method `command`should implement is `RunWithContext`?

Right! There are also another x methods need to be replaced. That's why I just introduced the `RunWithContext` but not replace them in one PR.

**Comment by 6543** [comment]
:point_up: that's what has to be done :)

`... replacing in all places in the src code with RunWithContext ...`

**Comment by 6543** [comment]
:point_up: that would be awesome

**Comment by lunny** [comment]
I think we can refactor them step by step. This PR is enough. Let's merge it at first.

File: modules/git/repo_stats.go [hunk 1]
```diff
@@ -95,22 +98,18 @@ func (repo *Repository) GetCodeActivityStats(fromTime time.Time, branch string)
 					continue
 				}
 				switch p {
-				case 1: // Separator
```
Comment (6543) [inline modules/git/repo_stats.go]: Why where these comments deleted?

**Comment by 6543** [comment]
@martinscholz83 thanks for the work, i only found one thing ☝️ and after it its good to go :)

File: modules/git/repo_stats.go (hunk 1 above)
Comment (martinscholz83) [reply modules/git/repo_stats.go]: Looks like while copy and paste the func the and then saved the file,  vscode linter removed or optimized these commands.

File: modules/git/repo_stats.go [hunk 2]
```diff
@@ -100,14 +103,10 @@ func (repo *Repository) GetCodeActivityStats(fromTime time.Time, branch string)
 					stats.CommitCount++
 				case 3: // Author
 					author = l
-				case 4: // E-mail
+				case 4:
```
Comment (6543) [inline modules/git/repo_stats.go]:
```suggestion
				case 4: // E-mail
```

File: modules/git/repo_stats.go [hunk 3]
```diff
@@ -67,12 +67,15 @@ func (repo *Repository) GetCodeActivityStats(fromTime time.Time, branch string)
...
-		nil, -1, repo.Path,
-		stdoutWriter, stderr, nil,
-		func(ctx context.Context, cancel context.CancelFunc) error {
+	err = NewCommand(repo.Ctx, args...).RunWithContext(&RunContext{
+		Env:     []string{},
+		Timeout: -1,
+		Dir:     repo.Path,
+		Stdout:  stdoutWriter,
+		Stderr:  stderr,
+		Stdin:   nil,
```
Comment (6543) [inline modules/git/repo_stats.go]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Comment by 6543** [comment]
hope you enjoined contributing :+1:

**Comment by martinscholz83** [comment]
🎉 Definitely! Thx for your help! To replace the rest of the `RunInDir` functions I started to work on.
```
		_, err := git.NewCommand(git.DefaultContext, "read-tree", "--empty").RunInDir(path)
```

These I got easily replaced with a regex script to
```
		err := git.NewCommand(git.DefaultContext, "read-tree", "--empty").RunWithContext(&RunContext{Dir: path, Timeout: -1,})
```
and then maybe check if `RunContext` can be resolved or if it has to be `git.RunContext`. But a lot of calls use the returned string (`stdout`) from `RunInDir` for later usage. So for example this
```
treeSha, err := git.NewCommand(git.DefaultContext, "write-tree").RunInDir(path)
treeSha = strings.TrimSpace(treeSha)
```
has to become this for example
```
stdout := new(bytes.Buffer)
err := git.NewCommand(git.DefaultContext, "write-tree").RunWithContext(&RunContext{Dir: path, Timeout: -1, Stdout: &stdout})
treeSha = strings.TrimSpace(stdout.String())
...
```


Is this what you have in mind?

**Comment by 6543** [comment]
☝️ yes 👍

**Pull request merged**